packages/http/src/validator/index.ts

```diff
@@ -37,7 +38,14 @@ const validateInput: ValidatorFn<IHttpOperation, IHttpRequest> = ({ resource, el
 const validateOutput: ValidatorFn<IHttpOperation, IHttpResponse> = ({ resource, element }) => {
   const results: IPrismDiagnostic[] = [];
   const mediaType = caseless(element.headers || {}).get('content-type');
-  const responseSpec = resource.responses && findOperationResponse(resource.responses, element.statusCode);
+  const responseSpec = resource.responses && toUndefined(findOperationResponse(resource.responses, element.statusCode));
+
+  if (!responseSpec) {
+    results.push({
+      message: 'Unable to match returned status code with those defined in spec',
+      severity: DiagnosticSeverity.Error,
+    });
+  }
 
   return results
```

Comment:
…and one when the responseSpec is defined; you can see indeed from the checks (`response&&responseSpec.contents`) that this piece of code is not executed if the response is not defined.

For this, we can use `fold`, which will accept two functions: one if Option has a content, one if the content does *not*. 

So we can rewrite this into

```ts
  return pipe(findOperationResponse(resource.responses, element.statusCode),
    fold(() => {
      results.push({
        message: 'Unable to match returned status code with those defined in spec',
        severity: DiagnosticSeverity.Error,
      });
      return results;

    }, responseSpec => {
      const mediaType = caseless(element.headers || {}).get('content-type');

      return results
        .concat(bodyValidator.validate(element.body, (responseSpec && responseSpec.contents) || [], mediaType))
        .concat(headersValidator.validate(element.headers || {}, (responseSpec && responseSpec.headers) || []));
    }))
```

Reply:
Two things here:
1) original code did not returned immediately when there was no responseSpec
1.1) now I think it should return, so the example provided by you is fine
2) theoretically if I want to preserve original flow, how should I model the situation where depending on existence of responseSpec we are pushing a diagnostic to array or not?

Reply:
@StefanDywersant 

In case you want to execute an action *only* when the `Option` is populated, you can either

1. Use a fold, with an empty handler for the "none" part.

`pipe(option, fold(()=>{}, value=> array.push(value))`

2. You can also use a map, which will only go through the "happy" part.

`pipe(option, map(value=>array.push(value))`

Reply:
Pushing to array creates a side-effect of invoking function, so I guess is not the FP-way.

Reply:
Generally speaking, you're right. The idea would be that map would return a new array.

However, it really depends; there are a lot of functional libraries that implement stuff in a non functional way, but as long it is isolated, it's fine anyway.

An example would be

```ts
function doSomething(a: number): Array<number> {
   const array: Array<number> = [];
   array.push(a);
   return a;
}
```

You can see this function is creating and pushing stuff in the array which is effectively non functional; however if you look at the code at the whole it's still ok as the creation and the destruction of the "non functional object" is confined in that function.

The code we're looking at fortunately for us is in this case. 

Does that make sense, clarify?

Reply:
The functional world is built on the foundation of an imperative one ;)
I just hoped that there are some helper functions provided by FP to solve it.

Thanks for the clarification!

Reply:
We can still refactor it — you're more than welcomed to refactor it if you want; I try to do this opportunistically when I see opportunities!

Reply:
Sure, just need to learn as much as I can about that FP. My goal is to leave every touched code as functional :)